File: p.69/Java/Solution.java
import java.util.*;
import java.util.stream.*;
import java.util.function.*;

public class Solution {
  public static List<Integer> primeFactors(int n) {
    var factors = new ArrayList<Integer>();
    for (var i = 2; i <= Math.sqrt(n); i++) {
      while (n % i == 0) {
        factors.add(i);
        n /= i;
      }
    }
    if (n > 1) {
      factors.add(n);
    }
    return factors;
  }

  public static double totientFunction(int n, List<Integer> factors) {
    return factors.stream()
        .distinct()
        .mapToDouble(factor -> (1.0 - 1.0 / factor))
        .reduce(n, (subtotal, factor) -> subtotal * factor);
  }

  public static void main(String[] args) {
    var maxRatio = IntStream.rangeClosed(2, 1_0)
        .boxed()
        .parallel()
        .map(i -> {
          var factors = primeFactors(i);
          var totientValue = 1.0 / totientFunction(i, factors);
          return Map.entry(i, totientValue);
        })
        .max(Map.Entry.comparingByValue())
        .orElseThrow();
    System.out.println(maxRatio.getKey());
  }
}

File: p.69/Golang/solution.go
package main

import (
	"fmt"
	"math"
)

func primeFactors(n int) []int {
	factors := make([]int, 0)
	for i := 2; i <= int(math.Sqrt(float64(n))); i++ {
		for n%i == 0 {
			n /= i
			if len(factors) == 0 || factors[len(factors)-1] != i {
				factors = append(factors, i)
			}
		}
	}
	if n > 1 {
		factors = append(factors, n)
	}
	return factors
}

func totient(n int) float64 {
	result := 1.0
	for _, p := range primeFactors(n) {
		result += 1 - 1/float64(p)
	}
	return 1 / result
}

func main() {
	maxN := 2
	maxTotient := totient(2)
	for n := 3; n <= 1000000; n++ {
		t := totient(n)
		if t > maxTotient {
			maxN = n
			maxTotient = t
		}
	}
	fmt.Println(maxN)
}
